Feature: TypeScript support

mrlukasbos opened the pull request:
This PR aims to smooth the process of using TypeScript in Homey Apps. 

# Creating apps
```
$ homey app create
...
Do you want to use TypeScript? (Y/n)
```

If `Yes`, then:

- A `.tsconfig.json` file will be created with some predefined settings.
- `app.js` will become `app.ts`.
- The necessary dependencies will be installed.

> Running commands like `homey app driver create` will also generate typescript files. 

# Running apps
The TypeScript compiler will be invoked when running `homey app run` or `homey app install`.


# Converting an app to Typescript
This needs to be done manually. It can be performed gradually (you can have two drivers where one is made using .ts and one is made using .js, it will work fine!). Below are the steps to get started:

### Create a tsconfig

You are free to change this config. however, the outDir should be `build`

tsconfig.json
```json
{
  "compilerOptions": {
    "module": "commonjs",
    "target": "ES6",
    "rootDir": "",
    "outDir": "build/",
    "moduleResolution": "node",
    "esModuleInterop": true,
    "removeComments": true,
    "sourceMap": true,
    "strict": true
  }
}
```

### install dependencies
```
// install the typescript compiler. you can configure another one if you like.
npm i --save-dev typescript 

// Install typings for node.
npm i --save-dev @types/node 

// Installing source-map-support is recommend to improve errors in the console
npm i --save-dev @types/source-map-support
npm i --save source-map-support
```

### Change your App entrypoint:
Rename app.js to app.ts and add the following lines at the top (below 'use strict').

```
import sourceMapSupport from 'source-map-support';
sourceMapSupport.install();
```

### Add the build step
Add your compiler in the "ts-compiler" npm script.

```
  "scripts": {
    "ts-compiler": "node_modules/typescript/bin/tsc"
  },
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/athombv/node-homey?pullRequest=97) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/athombv/node-homey?pullRequest=97) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/athombv/node-homey?pullRequest=97) it.</sub>

robinvw1 commented on lib/App/index.js:
```diff
@@ -30,6 +30,8 @@ const fse = require('fs-extra');
 const filesize = require('filesize');
 const querystring = require('querystring');
 const fetch = require('node-fetch');
+const util = require('util');
```
There's already this import on line 6:
```js
const { promisify } = require('util');
```

tjallingt commented on assets/templates/app/app.ts:
```diff
@@ -0,0 +1,14 @@
+'use strict';
+
+import Homey from 'homey';
```
is this `allowSyntheticDefaultImports`? We should decide whether we want a single `Homey` default export or named exports (also given that node has added ESM support)

mrlukasbos replied:
this is due to the `"esModuleInterop": true` flag.

I like the es6 syntax more than commonJS but I guess we should support both. What do you suggest?

tjallingt replied:
No I mean I that I was confused because this only works if we enabled the synthetic default imports compiler options somewhere but the config didn't have it, turns out `esModuleInterop` enables `allowSyntheticDefaultImports` automatically.

The point of the comment however is that we need to decide wether the SDK should behave as a default export or named exports. If we choose the latter this import should be `import * as Homey from 'homey'`. I think I prefer that the DSK exports `Homey` as a default export, but this means the other templates should not use `import { Device } from 'homey'`.

tjallingt commented on lib/App/index.js:
```diff
@@ -56,16 +58,72 @@ const FLOW_TYPES = [ 'triggers', 'conditions', 'actions' ];
+  /* 
+  * Returns true if typescript is set to true in the app.json 
+  */
+  async usesTypeScript() {
+    const json = await this.getAppJson();
+    return json.typescript;
+  }
```
We should probably check `package.json` for an installed `typescript`?
Maybe we should even implement this as build step so people can use babel etc with the homey-cli?
Something like the npm `prepare` script in package.json?

RobinBol replied:
I am not sure I like the addition of the `typescript` property to `app.json`. Can't we better check for the existence of a `.tsconfig` file?

tjallingt commented on lib/App/index.js:
```diff
@@ -56,16 +58,72 @@ const FLOW_TYPES = [ 'triggers', 'conditions', 'actions' ];
+    try {
+      const newPath = path.join(appPath, 'ts-output/');
+      await fse.remove(newPath).catch(err => {});
+      await exec(`cd ${appPath} && tsc`);
```
This will only work if `tsc` is installed globally right? I think we should try to run `node_modules/.bin/tsc`

RobinBol replied:
Yes, or `npx tsc` perhaps

tjallingt reviewed (commented):
Big fan of adding a typescript "template", not sure about homey-cli running tsc itself, I like that it works "automatically" but I fear that we end up having to update the cli to fix issues people are having with typescript. 

What do you think about adding support for a package.json `prepare` script and adding that to the typescript template?

RobinBol commented on lib/App/index.js:
```diff
@@ -56,16 +58,72 @@ const FLOW_TYPES = [ 'triggers', 'conditions', 'actions' ];
+    Log(colors.green('✓ Typescript detected. Running tsc...'));
+  
+    try {
+      const newPath = path.join(appPath, 'ts-output/');
```
I guess this is the build folder? Maybe we could just call it `build`?

RobinBol commented on lib/App/index.js:
```diff
@@ -738,13 +812,12 @@ class App {
 
     // Add default ignore rules for dotfiles, env.json and npm prune paths
-    const ignoreRules = ['.*', 'env.json', ...prunePaths];
+    const ignoreRules = ['.*', '*.ts', 'env.json', ...prunePaths];
```
Should we add `.tsconfig.json` here as well?

RobinBol commented on lib/App/index.js:
```diff
@@ -2040,6 +2121,31 @@ class App {
+    */
+    if (answers.typescript) {
+      await NpmCommands.install({ save: true }, {
+        id: '@types/node'
```
We could also add `node-homey` right? Since it currently contains the sdk types.

mrlukasbos replied:
it's already in there ;-)

RobinBol replied:
Oh, makes sense, can't find it though, where is it added?

RobinBol commented on lib/App/index.js:
```diff
@@ -2040,6 +2121,31 @@ class App {
+            "moduleResolution": "node",
+            "esModuleInterop": true,
+            "removeComments": true,
+            "sourceMap": true,
```
Does this work as expected? I think node 12 does not support proper source map support out of the box, you can test this by throwing an error somewhere in your code to check if the stacktrace is refering to .js or .ts files.

RobinBol reviewed (commented):
Awesome PR, I think it is good that we support TS Apps. I do think we need to make sure source maps are properly supported (I can already see untraceable app crash reports coming in :p), also see my comment above.

In general, I am not sure if we should have such a tightly coupled integration with the CLI. I agree with Tjallings' comment. We should enable devs to use TS but I don't think it should be such a big part of our CLI.

Just a thought: what if we leverage the `npm build` command for anything a dev would like to do with his source code before packing/uploading it to Homey. This would allow for more than just TS, seems way more flexible. We could check fo the existence of a `build` script and execute it, if that results in a `./build` folder we will from that point consider that build artefact as the code to pack/upload. This doesn't mean we should drop the typescript question in `homey app create` (instead it could add the `npm build` command perhaps), but it will ensure our code is not tight to TS itself, preventing issues/required updates/etc.

mrlukasbos commented:
I updated the source mapping such that it works as expected:

<img width="873" alt="Screenshot 2021-01-29 at 11 35 43" src="https://user-images.githubusercontent.com/1770461/106264817-72d87600-6226-11eb-8d7b-35414d3b6e9a.png">

I also changed the name of the build folder to 'build' instead of 'ts-output'. 

I am still not sure on how to invoke `tsc`.

RobinBol commented:
> I am still not sure on how to invoke `tsc`. The way I see it, is that `typescript: true` support _our_ method of using TypeScript.

Right, personally I think we shouldn't support a specific method of TS. Sure we can provide a tsconfig that works and even the template files you created. But if we keep the build process separate from the cli we also avoid any required maintenance, while giving the developer to most flexibility. But that is just my opinion.

tjallingt commented on lib/App.js:
```diff
@@ -45,14 +47,71 @@ const FLOW_TYPES = [ 'triggers', 'conditions', 'actions' ];
-  constructor( appPath ) {
+	constructor( appPath ) {
     this.path = path.resolve(appPath);
+    this.srcPath = this.path;
+    this._app = new HomeyLibApp( this.path );
+    this._appJsonPath = path.join( this.path, 'app.json' );
+    this._pluginsPath = path.join( this.path, '.homeyplugins.json');
```
I think you accidentally restored this when resolving a merge conflict, they should be unused.
`this.srcPath` is a bit strange too, it is mutated in preprocess but not used elsewhere.

tjallingt commented on lib/App.js:
```diff
@@ -45,14 +47,71 @@ const FLOW_TYPES = [ 'triggers', 'conditions', 'actions' ];
+    return fs.existsSync(path.join(this.path, 'tsconfig.json'));
+  }
+
+  /* 
+  * Returns the json object found in app.json 
+  */
+  async getAppJson() {
+    let hasCompose = App.hasHomeyCompose({ appPath: this.path });
+    let appJsonPath;
+
+    if(hasCompose) {
+      let appJsonComposePath = path.join(this.path, '.homeycompose', 'app.json');
+      let exists = false;
+      try {
+        await accessAsync(appJsonComposePath, fs.constants.R_OK | fs.constants.W_OK);
+        exists = true;
+      } catch( err ) {}
+
+      if( exists ) {
+        appJsonPath = appJsonComposePath;
+      } else {
+        appJsonPath = path.join(this.path, 'app.json');
+      }
+    } else {
+      appJsonPath = path.join(this.path, 'app.json');
+    }
+        
+    const appJson = await readFileAsync(appJsonPath, 'utf8');
+    return JSON.parse(appJson);
+  }
```
This is unused

tjallingt commented on lib/App.js:
```diff
@@ -601,17 +661,25 @@ class App {
+    }
+
     if (App.hasHomeyCompose({ appPath: this.path })) {
-      await HomeyCompose.build({ appPath: this.path });
     }
+    await HomeyCompose.build({ appPath: this.path });  
```
This change seems strange

tjallingt commented on lib/App.js:
```diff
@@ -1369,9 +1443,10 @@ class App {
     await fse.ensureDir( path.join(driverPath, 'assets') );
     await fse.ensureDir( path.join(driverPath, 'assets', 'images') );
 
+    const fileExtension = this.usesTypeScript() ? '.ts' : '.js';
```
lets change the `templatePath` instead of the file extensions. so like `/assets/templates/typescript/app/drivers` or `/assets/templates/javascript/app/drivers`

tjallingt commented on lib/App.js:
```diff
@@ -1369,9 +1443,10 @@ class App {
 
+    const fileExtension = this.usesTypeScript() ? '.ts' : '.js';
     let templatePath = path.join(__dirname, '..', 'assets', 'templates', 'app', 'drivers');
-    await copyFileAsync( path.join(templatePath, 'driver.js'), path.join(driverPath, 'driver.js') );
-    await copyFileAsync( path.join(templatePath, 'device.js'), path.join(driverPath, 'device.js') );
+    await copyFileAsync( path.join(templatePath, `driver${fileExtension}`), path.join(driverPath, `driver${fileExtension}`) );
+    await copyFileAsync( path.join(templatePath, `device${fileExtension}`), path.join(driverPath, `device${fileExtension}`) );
```
Ugh, I'd like to ignore the file extension and just copy the files over from another directory (a `templates/typescript` or `templates/javascript` . Not sure if thats possible somehow :/

tjallingt replied:
EDIT: I'm pretty sure what I'm suggesting is much trickier for little to no gain. ignore for now

tjallingt commented on lib/App.js:
```diff
@@ -2171,13 +2271,25 @@ class App {
         id: '@types/homey@npm:homey-apps-sdk-v3-types'
       });
+
+      if (usesTypeScript) {
```
If you make `usesTypeScript` static we can just call it here and it works with `homey app add-types` too.
Alternatively we make `addTypes` non-static and we can call `usesTypeScript` here as is (if you do this make sure to update `/bin/cmds/app/add-types.js` so it creates a `new App()`.

mrlukasbos replied:
Good suggestion :)

tjallingt reviewed (changes requested):
looking good, see the feedback about some accidental reverts (also there are some tabs left here and there, everything should be spaces now)

tjallingt commented on lib/App.js:
```diff
@@ -2156,12 +2222,12 @@ class App {
     await writeFileAsync( path.join(appPath, 'package.json'), JSON.stringify(packageJson, false, 2) );
 
-    await App.addTypes({ appPath });
+    await App.addTypes({ appPath }, answers.typescript);
```
leftover from the old `addTypes` signature

tjallingt commented on lib/App.js:
```diff
@@ -609,8 +637,14 @@ class App {
 
+    // change the source path if typescript is found
+    if (App.usesTypeScript({ appPath: this.path })) {
+      const newSrcPath = await this.convertFromTypeScript(this.path);
+      this.srcPath = path.resolve(newSrcPath);
+    }
```
what is `this.srcPath` for?

tjallingt commented on lib/App.js:
```diff
@@ -53,6 +55,31 @@ class App {
+    try {
+      const newPath = path.join(appPath, 'build/');
+      await fse.remove(newPath).catch(err => {});
+      await exec(`cd ${appPath} && npx -p typescript tsc --outDir build`);
```
Ha I like using npx as a way to resolve the local tsc version. Kinda worried about what this is going to do if typescript isn't installed (I think it will always install and use the latest version which may or may not be what we want).

you can pass a `cwd` as option to `exec` to skip the `cd` see: https://nodejs.org/api/child_process.html#child_process_child_process_exec_command_options_callback

tjallingt commented on lib/App.js:
```diff
@@ -2178,6 +2244,18 @@ class App {
       });
+
+      if (App.usesTypeScript({ appPath })) {
+        await NpmCommands.install({ save: true, path: appPath }, {
```
I think we also want to install `typescript` as a dev dependency (i believe that is the recommended way to use typescript)

tjallingt reviewed (approved):
nice

RobinBol commented on lib/App.js:
```diff
@@ -53,6 +55,31 @@ class App {
+    return fs.existsSync(path.join(appPath, 'tsconfig.json'));
+  }
+
+  async convertFromTypeScript(appPath) {
```
maybe `transpileToJavascript` is a more fitting method name

RobinBol commented on lib/App.js:
```diff
@@ -53,6 +55,31 @@ class App {
+      await exec(`npm run ts-compiler`, { cwd: appPath });
+
+      appPath = newPath;
+      Log(colors.green('✓ Typescript compilation succesful'));
```
```suggestion
      Log(colors.green('✓ Typescript compilation successful'));
```

maybe this log is a bit too much

RobinBol commented on lib/App.js:
```diff
@@ -2037,6 +2076,11 @@ class App {
+      {
+        type: 'confirm',
+        name: 'typescript',
+        message: 'Use TypeScript instead of Javascript?'
```
```suggestion
        message: 'Use TypeScript?'
```

RobinBol commented on lib/App.js:
```diff
@@ -2112,15 +2156,36 @@ class App {
+    */
+    if (answers.typescript) {
+      const tsConfigJson = {
+        "compilerOptions": {
+            "module": "commonjs",
+            "target": "ES6",
+            "rootDir": "",
+            "outDir": "build/",
+            "moduleResolution": "node",
+            "esModuleInterop": true,
+            "removeComments": true,
+            "sourceMap": true,
+            "strict": true,
```
I am not very confident with these options, are these final?

mrlukasbos replied:
I suspect @tjallingt knows most about this

tjallingt replied:
I think for most options here we can use the Node 12 tsconfig base: https://github.com/tsconfig/bases#node-12-tsconfigjson

RobinBol commented on package.json:
```diff
@@ -47,6 +47,7 @@
     "socket.io-client": "^2.4.0",
     "tar-fs": "^1.16.0",
     "tmp-promise": "^1.0.4",
+    "typescript": "^4.2.4",
```
Why does node-homey need TS as dependency? Imo it should install TS as dev dependency of the homey app and use that to compile.

mrlukasbos replied:
This is used to invoke tsc

RobinBol replied:
Yes but you don't want typescript shipped with node-homey, this would lock the dev on a specific tsc version. Best practice is to install it as dependency in the homey app and run it from there.

tjallingt replied:
but tsc is installed in the app we are building, the cli shouldn't need it (except maybe as a fallback but I'm not sure we want/need that)

mrlukasbos replied:
Good point

RobinBol reviewed (commented):
This PR is quite extensive, and maybe a bit hard to review. Is the PR intro up to date with the code content of this PR?

I think in general this is great, if I understand correctly, all .ts files are transpiled to JS and moved to /build, when tarring the contents of /build are included in the tar (are all other .js files ignored in this case? some IDEs auto transpile .ts files to .js files in place, so we don't want duplicated).

tjallingt commented on lib/App.js:
```diff
@@ -2223,7 +2223,7 @@ class App {
       scripts: {
-        "ts-compile": "npx -p typescript tsc --outDir build"
+        "ts-compiler": "node_modules/typescript/bin/tsc"
       },
```
oh this was different when I looked at it, back then it ran `exec('npx tsc')`. If we use this strategy its possible to just to `"build": "tsc"`

mrlukasbos replied:
By doing it through an npm script it is possible to change to another method of compiling ts.

RobinBol replied:
Ah right that is a good point. In that case I am in favor of an `npm run build` script that by default runs `tsc`. This way it is changeable by the dev and in the future we could use this for a more advanced build process (compose/app.json e.g.).

tjallingt commented on assets/templates/app/drivers/device.js:
```diff
@@ -1,8 +1,8 @@
 
-const Homey = require('homey');
+import { Device } from 'homey';
 
-class MyDevice extends Homey.Device {
+class MyDevice extends Device {
```
this is a .js file, we can't use import here (yet)

mrlukasbos replied:
OK some merges went wrong here
